pkg/backup/backupsink/sst_sink_key_writer.go

```
@@ -0,0 +1,310 @@
+			return nil
+		} else if isContiguous &&
+			(s.midRow || lastFile.EntryCounts.DataSize < fileSpanByteLimit) {
+			s.stats.spanGrows++
+			lastFile.Span.EndKey = newSpan.EndKey
+			return nil
+		}
+	}
+
+	// If we are resetting to a new span that is not contiguous with the
+	// last span, we need to ensure that we are not mid-row.
+	if lastFile != nil && !isContiguous && s.midRow {
+		if s.midRow {
+			return errors.AssertionFailedf(
+				"backup closed file ending mid-key in %v", lastFile.Span.EndKey,
+			)
+		}
+	}
+	s.flushedFiles = append(
+		s.flushedFiles,
+		backuppb.BackupManifest_File{
+			Span: newSpan,
+			Path: s.outName,
+		},
+	)
+	return nil
+}
+
+func (s *SSTSinkKeyWriter) Flush(ctx context.Context) error {
+	if len(s.flushedFiles) > 0 {
+		lastFile := s.flushedFiles[len(s.flushedFiles)-1]
+		// If WriteKey was used, it is possible that the last written key was
+		// mid-row, but because no keys were written after, s.midRow was not updated.
+		// To ensure this, we update midRow using the end of the span that the last
+		// key belonged to.
+		s.setMidRowForPrevKey(lastFile.Span.EndKey)
+	}
+	err := s.FileSSTSink.Flush(ctx)
+	s.prevKey = nil
+	return err
+}
+
+// TODO (kev-cao): Most of these `maybeFlushX` functions can be generalized to
+// other SSTSink implementations. When we get around to refactoring out the
+// other implementation, we should consider moving these functions into the base
+// FileSSTSink implementation.
+
+// maybeFlushOOOSpans checks if the new span's startKey precedes the last spans'
+// endKey or if the two spans do not share the same prefix based on the sink's
+// ElideMode. If either of these conditions are met, the current backup file is
+// flushed and a new one is created.
+func (s *SSTSinkKeyWriter) maybeFlushOOOSpans(ctx context.Context, newSpan roachpb.Span) error {
+	if len(s.flushedFiles) == 0 {
+		return nil
+	}
+	lastFile := &s.flushedFiles[len(s.flushedFiles)-1]
+	if newSpan.Equal(lastFile.Span) {
```

why would this ever happen?

---

Updated now with the assumption that having the same span should result in a flush.

---

To be clear, i don't think the user will ever pass a repeat span to the sink. We could throw an error.